Automated Smoketest: :x:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 Host: ip-10-20-0-83.ca-central-1.compute.internal
 GCC:  gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Git: 2.9.5,  CMake: 3.22.1,  cUrl: 7.67.0,  node.js: v16.13.1,  npm: 8.1.2
Sha: 9388404e366b8673ba312c5692ebcd9154ade919
Containerized:False
Number of Error(s): 8
warn(226) A package without a profile could throw errors or warnings. package: dojo-util
warn(226) A package without a profile could throw errors or warnings. package: dojo-util
\terrors: 0
/home/centos/smoketest/PR-15891/HPCC-Platform/esp/services/ldapenvironment/ldapenvironmentService.cpp:283:83: error: expected \xe2\x80\x98,\xe2\x80\x99 or \xe2\x80\x98;\xe2\x80\x99 before \xe2\x80\x98)\xe2\x80\x99 token
/home/centos/smoketest/PR-15891/HPCC-Platform/esp/services/ldapenvironment/ldapenvironmentService.cpp:361:79: error: expected \xe2\x80\x98,\xe2\x80\x99 or \xe2\x80\x98;\xe2\x80\x99 before \xe2\x80\x98)\xe2\x80\x99 token
make[2]: *** [esp/services/ldapenvironment/CMakeFiles/ldapenvironment.dir/build.make:90: esp/services/ldapenvironment/CMakeFiles/ldapenvironment.dir/ldapenvironmentService.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:13393: esp/services/ldapenvironment/CMakeFiles/ldapenvironment.dir/all] Error 2
make: *** [Makefile:166: all] Error 2
Get ECLWatch build result
ReportTimes.
Finish at: 2022-03-14_15-18-57
time logs started
Prep time   :    23 sec (00:00:23)
Build time  :   289 sec (00:04:49)
Package time:     0 sec (00:00:00)
Install time:     0 sec (00:00:00)
Start time  :     0 sec (00:00:00)
Test time   :     0 sec (00:00:00)
Stop time   :     0 sec (00:00:00)
Summary     :   312 sec (00:05:12)
time logs end

[Build error:](https://gist.github.com/a0a1355c0103be7185c439852b415012)
Build log [Build log](https://gist.github.com/HPCCSmoketest/a0a1355c0103be7185c439852b415012#file-result-22-03-14-15-13-28-log)

--- a/esp/applications/common/ldap/ldap.yaml
+++ b/esp/applications/common/ldap/ldap.yaml
@@ -9,7 +9,7 @@ ldap:
   maxConnections: 10
   passwordExpirationWarningDays: 10
   cacheTimeout: 5
-  ldapTimeoutSecs: 131
+  ldapTimeoutSecs: 30
   sharedCache: true
   filesBasedn: ou=files,ou=ecl
   groupsBasedn: ou=groups,ou=ecl

--- a/esp/scm/ldapenvironment.ecm
+++ b/esp/scm/ldapenvironment.ecm
@@ -44,7 +44,7 @@ ESPenum OUMode : int
 ESPrequest LDAPCreateEnvironmentRequest
 {
     bool                                              CreateLDAPEnvironment(false);
-    bool                                              CreateK8sSecret(false);
+    bool                                              CreateK8sSecrets(false);
     [label("Environment Name"), cols(20)] string      EnvName;
     [label("Environment Owner"), cols(20)] string     EnvOwnerName;
     [label("Description"), cols(60)] string           EnvDescription;
@@ -68,6 +68,8 @@ ESPrequest LDAPCreateEnvironmentRequest
 ESPresponse [exceptions_inline] LDAPCreateEnvironmentResponse
 {
     string  Notes;
+    string  LDAPAdminUsername;
+    string  LDAPAdminPassword;
     string  HPCCAdminUsername;
     string  HPCCAdminPassword;
     string  LDAPHelm;

--- a/esp/services/ldapenvironment/ldapenvironmentService.cpp
+++ b/esp/services/ldapenvironment/ldapenvironmentService.cpp
@@ -108,6 +108,59 @@ const char * CldapenvironmentEx::formatOUname(StringBuffer &ou, const char * env
     return ou.str();
 }
 
+bool CldapenvironmentEx::changePermissions(const char * ou, const char * userFQDN, SecAccessFlags allows, SecAccessFlags denies)
+{
+    // Given an OU such as "ou=BocaInsurance,ou=hpcc,dc=myldap,dc=com", the
+    // 'rName' is 'BocaInsurance' and the baseDN is 'ou=hpcc,dc=myldap,dc=com"
+    StringBuffer baseDN;
+    StringBuffer rName;
+
+    size_t finger = 0;
+    while ('=' != ou[finger++]);//skip over "ou=" or "cn="
+    while (',' != ou[finger])
+        rName.append( (char)ou[finger++] );
+    baseDN.append( ou + finger + 1 );
+
+    CPermissionAction action;
+    action.m_action = "update";
+    action.m_basedn = baseDN;
+    action.m_rname =  rName;
+    action.m_rtype = RT_DEFAULT;
+    action.m_account_name = userFQDN;//fully qualified user DN
+    action.m_account_type = USER_ACT;
+    action.m_allows = allows;
+    action.m_denies = denies;
+    DBGLOG("Setting (%d,%d) permissions for rName %s, baseDN %s, user %s", (int)allows, (int)denies, rName.str(), baseDN.str(), userFQDN);
+    bool ok;
+    try
+    {
+        ok = secmgr->changePermission(action);
+    }
+    catch (...)
+    {
+    }
+    return ok;
+}
+
+bool CldapenvironmentEx::createSecret(const char * secretName, const char * username, const char * pwd, StringBuffer & notes)
+{
+    VStringBuffer cmdLineSafe("kubectl create secret generic %s --from-literal=username=%s --from-literal=password=", secretName, username);
+    VStringBuffer cmdLine("%s%s", cmdLineSafe.str(), pwd);
+    try
+    {
+        DBGLOG("\nExecuting '%s'\n", cmdLineSafe.str());
+        long unsigned int runcode;
+        bool success = invoke_program(cmdLine.str(), runcode) && (runcode==0);
+        if (!success)
+            notes.appendf("\nError executing '%s', please run manually before executing helm install", cmdLineSafe.str());
+    }
+    catch (...)
+    {
+        notes.appendf("\nException executing '%s', please run manually before executing helm install", cmdLineSafe.str());
+    }
+    return true;
+}
+
 
 bool CldapenvironmentEx::onLDAPCreateEnvironment(IEspContext &context, IEspLDAPCreateEnvironmentRequest &req, IEspLDAPCreateEnvironmentResponse &resp)
 {
@@ -142,7 +195,7 @@ bool CldapenvironmentEx::onLDAPCreateEnvironment(IEspContext &context, IEspLDAPC
         if (!req.getWorkunitsMode() == COUMode_CreateCustom && isEmptyString(req.getCustomWorkunitsBaseDN()))
             throw MakeStringException(-1, "CustomWorkunitsBaseDN must be specified (ex. 'ou=workunits,ou=hpcc,dc=myldap,dc=com')");
 
-        // Create OU names
+        // Create OU string names
 
         StringBuffer respFilesBaseDN, respGroupsBaseDN, respUsersBaseDN, respResourcesBaseDN, respWorkunitsBaseDN;
         formatOUname(respFilesBaseDN, req.getEnvName(), req.getFilesMode(), sharedFilesBaseDN.str(), req.getCustomFilesBaseDN(), "ou=files");
@@ -159,77 +212,205 @@ bool CldapenvironmentEx::onLDAPCreateEnvironment(IEspContext &context, IEspLDAPC
         }
 
         //Create LDAP resources OU Hierarchy
-
+        VStringBuffer description("%s, created by %s", req.getEnvDescription(), req.getEnvOwnerName());
         if (req.getCreateLDAPEnvironment())
         {
-            VStringBuffer description("%s, created by %s", req.getEnvDescription(), req.getEnvOwnerName());
             //Note that ESP will also try to create these OUs on startup
-            try { secmgr->createLdapBasedn(nullptr, respFilesBaseDN.str(), PT_ADMINISTRATORS_ONLY, description.str()); }
-            catch(...) { notes.appendf("\nError creating '%s'", respFilesBaseDN.str()); }
-            try { secmgr->createLdapBasedn(nullptr, respGroupsBaseDN.str(), PT_ADMINISTRATORS_ONLY, description.str()); }
-            catch(...) { notes.appendf("\nError creating '%s'", respGroupsBaseDN.str()); }
-            try { secmgr->createLdapBasedn(nullptr, respUsersBaseDN.str(), PT_ADMINISTRATORS_ONLY, description.str()); }
-            catch(...) { notes.appendf("\nError creating '%s'", respUsersBaseDN.str()); }
-            try { secmgr->createLdapBasedn(nullptr, respResourcesBaseDN.str(), PT_ADMINISTRATORS_ONLY, description.str()); }
-            catch(...) { notes.appendf("\nError creating '%s'", respResourcesBaseDN.str()); }
-            try { secmgr->createLdapBasedn(nullptr, respWorkunitsBaseDN.str(), PT_ADMINISTRATORS_ONLY, description.str()); }
-            catch(...) { notes.appendf("\nError creating '%s'", respWorkunitsBaseDN.str()); }
+            try
+            {
+                 secmgr->createLdapBasedn(nullptr, respFilesBaseDN.str(), PT_ADMINISTRATORS_ONLY, description.str());
+            }
+            catch(...)
+            {
+                  notes.appendf("\nNon Fatal Error creating '%s'", respFilesBaseDN.str());
+            }
+            try
+            {
+                 secmgr->createLdapBasedn(nullptr, respGroupsBaseDN.str(), PT_ADMINISTRATORS_ONLY, description.str());
+            }
+            catch(...)
+            {
+                notes.appendf("\nNon Fatal Error creating '%s'", respGroupsBaseDN.str());
+            }
+            try
+            {
+                secmgr->createLdapBasedn(nullptr, respUsersBaseDN.str(), PT_ADMINISTRATORS_ONLY, description.str());
+            }
+            catch(...)
+            {
+                notes.appendf("\nNon Fatal Error creating '%s'", respUsersBaseDN.str());
+            }
+            try
+            {
+                secmgr->createLdapBasedn(nullptr, respResourcesBaseDN.str(), PT_ADMINISTRATORS_ONLY, description.str());
+            }
+            catch(...)
+            {
+                notes.appendf("\nNon Fatal Error creating '%s'", respResourcesBaseDN.str());
+            }
+            try
+            {
+                secmgr->createLdapBasedn(nullptr, respWorkunitsBaseDN.str(), PT_ADMINISTRATORS_ONLY, description.str());
+            }
+            catch(...)
+            {
+                 notes.appendf("\nNon Fatal Error creating '%s'", respWorkunitsBaseDN.str());
+            }
+
+            //Create HPCCAdmins Group
+            try
+            {
+                secmgr->addGroup(adminGroupName.str(), nullptr, description.str(), respGroupsBaseDN.str());
+            }
+            catch(...)
+            {
+                notes.appendf("\nNon Fatal Error creating '%s,%s'", adminGroupName.str(), respGroupsBaseDN.str());
+            }
         }
         
         //----------------------------------
         // Create HPCC Admin Username/password.
         // Attempt to create the Kubernetes secret for that user
-        // NOTE!  HPCCAdmin LDAP group and HPCCAdmin LDAP user will be
-        // created by ESP when started in Admin mode
-        //----------------------------------        
-        VStringBuffer respHPCCAdminUser("admin_%s", req.getEnvName());
-
-        VStringBuffer  hpccAdminK8sSecretName("hpccadminsecret-%s", req.getEnvName());
-        hpccAdminK8sSecretName.toLowerCase();
-
+        //----------------------------------
+        VStringBuffer respHPCCAdminUser("adminHPCC_%s", req.getEnvName());
         StringBuffer  respHPCCAdminPwd;
         generatePassword(respHPCCAdminPwd, 10);//jutil.hpp
 
-        if (req.getCreateK8sSecret())
+        if (req.getCreateLDAPEnvironment())
         {
-            VStringBuffer cmdLineSafe("kubectl create secret generic %s --from-literal=username=%s --from-literal=password=", hpccAdminK8sSecretName.str(), respHPCCAdminUser.str());
-            VStringBuffer cmdLine("%s%s", cmdLineSafe.str(), respHPCCAdminPwd.str());
-            try
+            //Create the HPCCAdmin user
             {
-                DBGLOG("\nExecuting '%s'\n", cmdLineSafe.str());
-                long unsigned int runcode;
-                bool success = invoke_program(cmdLine.str(), runcode) && (runcode==0);
-                if (!success)
-                    notes.appendf("\nError executing '%s', please run manually before executing helm install", cmdLineSafe.str());
+                Owned<ISecUser> user = secmgr->createUser(respHPCCAdminUser.str());
+                user->credentials().setPassword(respHPCCAdminPwd.str());
+                try
+                {
+                    secmgr->addUser(*user.get(), respUsersBaseDN.str());
+                }
+                catch(...)
+                {
+                    notes.appendf("\nNon Fatal Error creating '%s'", respHPCCAdminUser.str());
+                }
             }
-            catch (...)
+
+            const char * userPrefix = secmgr->getLdapServerType() == ACTIVE_DIRECTORY ? "cn=" : "uid=";
+            const char * resPrefix = secmgr->getLdapServerType() == ACTIVE_DIRECTORY ? "cn=" : "ou=";
+
+            //Add HPCCAdmin user to HPCCAdmins group
             {
-                notes.appendf("\nException executing '%s', please run manually before executing helm install", cmdLineSafe.str());
+                VStringBuffer adminGrpOU("cn=%s,%s", adminGroupName.str(), respGroupsBaseDN.str());
+                VStringBuffer adminUsr("%s%s,%s", userPrefix, respHPCCAdminUser.str(), respUsersBaseDN.str());
+                try
+                {
+                    secmgr->changeGroupMember("add", adminGrpOU.str(), adminUsr.str());
+                }
+                catch(...)
+                {
+                    notes.appendf("\nNon Fatal Error adding '%s' to '%s'", adminUsr.str(), adminGrpOU.str());
+                }
+            }
+
+            //Grant SmcAccess to HPCCAdmins group
+            {
+                VStringBuffer smcAccess("%sSmcAccess,%s", resPrefix, respResourcesBaseDN.str());
+                try
+                {
+                    secmgr->createLdapBasedn(nullptr, smcAccess.str(), PT_ADMINISTRATORS_ONLY, description.str());
+                }
+                catch(...)
+                {
+                    notes.appendf("\nNon Fatal Error creating '%s'", smcAccess.str());
+                }
+
+                CPermissionAction action;
+                action.m_action = "update";
+                action.m_basedn = respResourcesBaseDN.str();
+                action.m_rname = "SmcAccess";
+                action.m_rtype = RT_SERVICE;
+                action.m_account_name = adminGroupName.str();
+                action.m_account_type = GROUP_ACT;
+                action.m_allows = SecAccess_Full;
+                action.m_denies = 0;
+                try
+                {
+                    secmgr->changePermission(action);
+                }
+                catch(...)
+                {
+                    notes.appendf("\nNon Fatal Error setting '%s' permission for '%s'", smcAccess.str(), adminGroupName.str());
+                }
             }
         }
+
+        //Create the secret
+        VStringBuffer  respHPCCAdminK8sSecretName("hpcc-admin-%s", req.getEnvName());
+        respHPCCAdminK8sSecretName.toLowerCase();
+        if (req.getCreateK8sSecrets())
+            createSecret(respHPCCAdminK8sSecretName.str(), respHPCCAdminUser.str(), respHPCCAdminPwd.str(), notes);
+
+        //----------------------------------
+        // Create LDAP Admin Username/password.
+        // Attempt to create the Kubernetes secret for that user
+        //----------------------------------
+        VStringBuffer respLDAPAdminUser("adminLDAP_%s", req.getEnvName());
+        StringBuffer  respLDAPAdminPwd;
+        generatePassword(respLDAPAdminPwd, 10);//jutil.hpp
+
+        if (req.getCreateLDAPEnvironment())
+        {
+            //Create the user
+            Owned<ISecUser> user = secmgr->createUser(respLDAPAdminUser.str());
+            user->credentials().setPassword(respLDAPAdminPwd.str());
+            try {secmgr->addUser(*user.get(), respUsersBaseDN.str()); }
+            catch(...) { notes.appendf("\nNon Fatal Error creating '%s'", respLDAPAdminUser.str()); }
+
+            //Add LDAP R/W permissions for LDAPAdmin user
+            //Only grant access to root of new environment (ex  ou=BocaInsurance,ou=hpcc,dc=myldap,dc=com)
+            VStringBuffer ldapAdminFQDN("uid=%s,%s", respLDAPAdminUser.str(), respUsersBaseDN.str());//@@
+            if (!changePermissions(envOU.str(), ldapAdminFQDN.str(), SecAccess_Full, SecAccess_None))
+                notes.appendf("\nNon Fatal Error setting LDAPAdmin permission for %s'", envOU.str());
+        }
+
+        //Create the secret
+        VStringBuffer respLDAPAdminK8sSecretName("ldap-admin-%s", req.getEnvName());
+        respLDAPAdminK8sSecretName.toLowerCase();
+        if (req.getCreateK8sSecrets())
+            createSecret(respLDAPAdminK8sSecretName.str(), respLDAPAdminUser.str(), respLDAPAdminPwd.str(), notes);
 
         //----------------------------------
         // Set response
         //----------------------------------
+        VStringBuffer ldapcredskey("ldapcredskey-%s", req.getEnvName());
+        VStringBuffer hpcccredskey("hpcccredskey-%s", req.getEnvName());
+        ldapcredskey.toLowerCase();
+        hpcccredskey.toLowerCase();
+
+        resp.setLDAPAdminUsername(respLDAPAdminUser.str());
+        resp.setLDAPAdminPassword(respLDAPAdminPwd.str());
         resp.setHPCCAdminUsername(respHPCCAdminUser.str());
         resp.setHPCCAdminPassword(respHPCCAdminPwd.str());
-        VStringBuffer ldapHelm("\n"
+        VStringBuffer ldapHelm("\n\n"
                                "secrets:\n"
                                "  authn:\n"
-                               "    myhpcccreds: %s\n\n"
+                               "    %s: %s\n"
+                               "    %s: %s\n\n"
                                "esp:\n"
                                "- name: eclwatch\n"
                                "  auth: ldap\n"
                                "  ldap:\n"
                                "    adminGroupName: %s\n"
-                               "    hpccAdminSecretKey: myhpcccreds\n"
+                               "    ldapAdminSecretKey: %s\n"
+                               "    hpccAdminSecretKey: %s\n"
                                "    filesBasedn: %s\n"
                                "    groupsBasedn: %s\n"
                                "    usersBasedn: %s\n"
                                "    resourcesBasedn: %s\n"
-                               "    workunitsBasedn: %s\n",
-                               hpccAdminK8sSecretName.str(), adminGroupName.str(), respFilesBaseDN.str(), respGroupsBaseDN.str(),
-                               respUsersBaseDN.str(), respResourcesBaseDN.str(), respWorkunitsBaseDN.str());
+                               "    workunitsBasedn: %s\n"
+                               "    systemBasedn: %s\n\n",
+                               ldapcredskey.str(), respLDAPAdminK8sSecretName.str(),
+                               hpcccredskey.str(), respHPCCAdminK8sSecretName.str(),
+                               adminGroupName.str(),
+                               ldapcredskey.str(), hpcccredskey.str(),
+                               respFilesBaseDN.str(), respGroupsBaseDN.str(), respUsersBaseDN.str(), respResourcesBaseDN.str(), respWorkunitsBaseDN.str(), respUsersBaseDN.str());
         resp.setLDAPHelm(ldapHelm.str());
         resp.setNotes(notes.str());
     }

--- a/esp/services/ldapenvironment/ldapenvironmentService.hpp
+++ b/esp/services/ldapenvironment/ldapenvironmentService.hpp
@@ -32,6 +32,8 @@ private:
     StringBuffer    adminGroupName;
 
     const char * formatOUname(StringBuffer &ou, const char * envName, int mode, const char * sharedOU, const char * reqBaseDN, const char * privateOU);
+    bool createSecret(const char * secretName, const char * username, const char * pwd, StringBuffer & notes);
+    bool changePermissions(const char * ou, const char * userFQDN, SecAccessFlags allows, SecAccessFlags denies);
 
 public:
     IMPLEMENT_IINTERFACE;

--- a/system/jlib/jutil.cpp
+++ b/system/jlib/jutil.cpp
@@ -3353,25 +3353,29 @@ const char * generatePassword(StringBuffer &pwd, int pwdLen)
 
     //Ensures each character group used at least once
     srand(time(0));
-    pwd.append(alphaUC[rand() % sizeof(alphaUC)]);
-    pwd.append(alphaLC[rand() % sizeof(alphaLC)]);
-    pwd.append(numeric[rand() % sizeof(numeric)]);
-    pwd.append(symbol[rand() % sizeof(symbol)]);
+    pwd.append(alphaUC[rand() % (sizeof(alphaUC) - 1)]);
+    pwd.append(alphaLC[rand() % (sizeof(alphaLC) - 1)]);
+    pwd.append(numeric[rand() % (sizeof(numeric) - 1)]);
+    pwd.append(symbol[rand() % (sizeof(symbol) - 1)]);
 
     for (int i = 4; i < pwdLen; i++)
     {
         switch(rand() % NUM_GROUPS)//select a random character group
         {
         case 0:
-            pwd.append(alphaUC[rand() % sizeof(alphaUC)]);  break;
+            pwd.append(alphaUC[rand() % (sizeof(alphaUC) - 1)]);  break;
         case 1:
-            pwd.append(alphaLC[rand() % sizeof(alphaLC)]);  break;
+            pwd.append(alphaLC[rand() % (sizeof(alphaLC) - 1)]);  break;
         case 2:
-            pwd.append(numeric[rand() % sizeof(numeric)]);  break;
+            pwd.append(numeric[rand() % (sizeof(numeric) - 1)]);  break;
         case 3:
-            pwd.append(symbol[rand() % sizeof(symbol)]);  break;
+            pwd.append(symbol[rand() % (sizeof(symbol) - 1)]);  break;
         }
     }
+#ifdef _DEBUG
+    if (pwd.length() != (size32_t)pwdLen)
+        throw makeStringException(0, "Generated Passwords wrong length!");
+#endif
     return pwd.str();
 }
 

--- a/system/security/LdapSecurity/ldapconnection.cpp
+++ b/system/security/LdapSecurity/ldapconnection.cpp
@@ -304,6 +304,9 @@ private:
         DBGLOG("m_view_basedn = %s", m_view_basedn.str());
         DBGLOG("m_workunitscope_basedn = %s", m_workunitscope_basedn.str());
 
+        DBGLOG("m_HPCCAdminUser_username = %s", m_HPCCAdminUser_username.str());
+//        DBGLOG("m_HPCCAdminUser_password = %s", m_HPCCAdminUser_password.str());
+
         DBGLOG("m_sysuser = %s", m_sysuser.str());
         DBGLOG("m_sysuser_dn = %s", m_sysuser_dn.str());
         DBGLOG("m_sysuser_commonname = %s", m_sysuser_commonname.str());
@@ -449,7 +452,10 @@ public:
 
             //Guesstimate system user baseDN based on config settings. It will be used if anonymous bind fails
             StringBuffer sysUserDN;
-            sysUserDN.append("cn=").append(m_sysuser_commonname.str()).append(",").append(sysBasedn.str());
+            if(m_serverType == OPEN_LDAP)
+                sysUserDN.append("uid=").append(m_sysuser_commonname).append(",").append(sysBasedn);
+            else
+                sysUserDN.append("cn=").append(m_sysuser_commonname).append(",").append(sysBasedn);
 
             for(int retries = 0; retries <= LDAPSEC_MAX_RETRIES; retries++)
             {
@@ -459,7 +465,7 @@ public:
                 sleep(LDAPSEC_RETRY_WAIT);
                 if(retries < LDAPSEC_MAX_RETRIES)
                 {
-                    DBGLOG("Server %s temporarily unreachable, retrying...", hostbuf.str());
+                    DBGLOG("LDAP AD Server %s temporarily unreachable for user %s, retrying...", hostbuf.str(), sysUserDN.str());
                 }
             }
             if (rc != LDAP_SUCCESS)
@@ -649,10 +655,12 @@ public:
                 else
                 {
                     if (nullptr == strchr(m_sysuser_commonname.str(), '='))
-                        m_sysuser_dn.append("uid=").append(m_sysuser_commonname.str()).append(",").append(m_sysuser_basedn.str()).append(",").append(m_basedn.str());
+                        m_sysuser_dn.append("uid=").append(m_sysuser_commonname.str()).append(",").append(m_sysuser_basedn.str());
                     else
                         m_sysuser_dn.append(m_sysuser_commonname.str());//includes FQDN prefix, use as is (likely from initldap)
                 }
+                if (!strstr(m_sysuser_dn.str(), ",dc="))
+                    m_sysuser_dn.append(",").append(m_basedn.str());
             }
         }
 
@@ -747,6 +755,11 @@ public:
     virtual const char* getUserBasedn()
     {
         return m_user_basedn.str();
+    }
+
+    virtual void setUserBasedn(const char * basedn)
+    {
+        m_user_basedn.set(basedn);
     }
 
     virtual const char* getGroupBasedn()
@@ -1687,7 +1700,7 @@ public:
 
                     //Create HPCC Admin user
                     const char * pUser = m_ldapconfig->getHPCCAdminUser_username();
-                    if (pUser)
+                    if (!isEmptyString(pUser))
                     {
                         DBGLOG("Creating HPCC Admin user %s", pUser);
                         Owned<ISecUser> user = new CLdapSecUser(pUser, nullptr);
@@ -4185,13 +4198,17 @@ public:
 
         if(m_ldapconfig->getServerType() != ACTIVE_DIRECTORY)
         {
-            StringBuffer act_dn;
-            if(action.m_account_type == GROUP_ACT)
-                getGroupDN(action.m_account_name.str(), act_dn);
-            else
-                getUserDN(action.m_account_name.str(), act_dn);
-            
-            action.m_account_name.clear().append(act_dn.str());
+            //if not already fully qualified
+            if (0 == strstr(action.m_account_name, "dc=") &&
+                0 == strstr(action.m_account_name, "ou="))
+            {
+                StringBuffer act_dn;
+                if(action.m_account_type == GROUP_ACT)
+                    getGroupDN(action.m_account_name.str(), act_dn);
+                else
+                    getUserDN(action.m_account_name.str(), act_dn);
+                action.m_account_name.set(act_dn.str());
+            }
         }
 
         Owned<CSecurityDescriptor> newsd = m_pp->changePermission(sd.get(), action);
@@ -6227,6 +6244,15 @@ private:
 
     }
 
+    //Add new user to the given base DN
+    virtual bool addUser(ISecUser & user, const char* basedn)
+    {
+        StringBuffer prevBaseDN(m_ldapconfig->getUserBasedn());
+        m_ldapconfig->setUserBasedn(basedn);
+        bool rc = addUser(user);
+        m_ldapconfig->setUserBasedn(prevBaseDN.str());
+        return rc;
+    }
 
     virtual bool addUser(ISecUser& user)
     {

--- a/system/security/LdapSecurity/ldapconnection.hpp
+++ b/system/security/LdapSecurity/ldapconnection.hpp
@@ -213,6 +213,7 @@ interface ILdapConfig : extends IInterface
     virtual void setResourceBasedn(const char* rbasedn, SecResourceType rtype = RT_DEFAULT) = 0;
     virtual int getLdapTimeout() = 0;
     virtual bool isAzureAD() = 0;
+    virtual void setUserBasedn(const char * basedn) = 0;
 };
 
 
@@ -301,6 +302,7 @@ interface ILdapClient : extends IInterface
         ResourcePermissionField* sortOrder, const unsigned pageStartFrom, const unsigned pageSize, unsigned* total, __int64* cacheHint) = 0;
     virtual bool changePermission(CPermissionAction& action) = 0;
     virtual void changeUserGroup(const char* action, const char* username, const char* groupname, const char * groupDN=nullptr) = 0;
+    virtual void changeGroupMember(const char* action, const char* groupdn, const char* userdn) = 0;
     virtual bool deleteUser(ISecUser* user) = 0;
     virtual void addGroup(const char* groupname, const char * groupOwner, const char * groupDesc) = 0;
     virtual void addGroup(const char* groupname, const char * groupOwner, const char * groupDesc, const char* basedn) = 0;
@@ -335,6 +337,7 @@ interface ILdapClient : extends IInterface
     virtual bool userInView(const char * user, const char* viewName) = 0;
     virtual void createLdapBasedn(ISecUser* user, const char* basedn, SecPermissionType ptype, const char* description) = 0;
     virtual const bool organizationalUnitExists(const char * ou) const = 0;
+    virtual bool addUser(ISecUser & user, const char* basedn) = 0;
 };
 
 ILdapClient* createLdapClient(IPropertyTree* cfg);

--- a/system/security/LdapSecurity/ldapsecurity.cpp
+++ b/system/security/LdapSecurity/ldapsecurity.cpp
@@ -1429,6 +1429,11 @@ void CLdapSecManager::changeUserGroup(const char* action, const char* username, 
     m_ldap_client->changeUserGroup(action, username, groupname);
 }
 
+void CLdapSecManager::changeGroupMember(const char* action, const char* groupdn, const char* userdn)
+{
+    m_ldap_client->changeGroupMember(action, groupdn, userdn);
+}
+
 bool CLdapSecManager::deleteUser(ISecUser* user)
 {
     return m_ldap_client->deleteUser(user);
@@ -1628,6 +1633,11 @@ void CLdapSecManager::createLdapBasedn(ISecUser* user, const char* basedn, SecPe
 const bool CLdapSecManager::organizationalUnitExists(const char * ou) const
 {
     return m_ldap_client->organizationalUnitExists(ou);
+}
+
+bool CLdapSecManager::addUser(ISecUser & user, const char* basedn)
+{
+    return m_ldap_client->addUser(user, basedn);
 }
 
 extern "C"

--- a/system/security/LdapSecurity/ldapsecurity.ipp
+++ b/system/security/LdapSecurity/ldapsecurity.ipp
@@ -427,6 +427,7 @@ public:
     virtual void getGroups(const char* username, StringArray & groups);
     virtual bool changePermission(CPermissionAction& action);
     virtual void changeUserGroup(const char* action, const char* username, const char* groupname);
+    virtual void changeGroupMember(const char* action, const char* groupdn, const char* userdn);
     virtual bool deleteUser(ISecUser* user);
     virtual void addGroup(const char* groupname, const char * groupOwner, const char * groupDesc);
     virtual void addGroup(const char* groupname, const char * groupOwner, const char * groupDesc, const char* basedn);
@@ -492,6 +493,7 @@ public:
     bool removeResources(ISecUser& sec_user, ISecResourceList * resources, IEspSecureContext* secureContext = nullptr) override { return false; }
     virtual void createLdapBasedn(ISecUser* user, const char* basedn, SecPermissionType ptype, const char* description);
     virtual const bool organizationalUnitExists(const char * ou) const;
+    virtual bool addUser(ISecUser & user, const char* basedn);
     
     //Data View related interfaces
     virtual void createView(const char * viewName, const char * viewDescription);

--- a/system/security/LdapSecurity/ldaputils.cpp
+++ b/system/security/LdapSecurity/ldaputils.cpp
@@ -231,7 +231,7 @@ LDAP* LdapUtils::ldapInitAndSimpleBind(const char* ldapserver, const char* userD
     *err = LdapSimpleBind(ld, timeout, (char*)userDN, (char*)pwd);
     if (*err != LDAP_SUCCESS)
     {
-        DBGLOG("LdapSimpleBind error (%d) - %s for admin user %s", *err, ldap_err2string(*err), isEmptyString(userDN) ? "NULL" : userDN);
+        DBGLOG("LdapSimpleBind error (%d) - '%s' for admin user %s", *err, ldap_err2string(*err), isEmptyString(userDN) ? "NULL" : userDN);
         if (!isEmptyString(userDN))
             DBGLOG("Please make sure your LDAP configuration 'systemBasedn' contains the complete path, including the complete 'dc=domainComponent'");
         return nullptr;

